### DISCUSSION TITLE
Feedback on the new Bootstrap 5/Vue 3 look & feel

### DISCUSSION THREAD

**Discussion by Moonbase59** [AzuraCast/AzuraCast#6406] [General]
First, let me say I appreciate all the work "behind the scenes" that has gone into this, and the changes make a lot of sense.

Here are my first critical impressions, based on the Rolling Release #[40155bd](https://github.com/AzuraCast/AzuraCast/commit/40155bd27725824e61bf76aa5532e11c782544ca) (2023-07-09 4:29) update:

- Fonts are bigger now, you can’t get as much onto a 2/3 wide browser on a 1920x1080 desktop.
- Everything has much more contrasting colors, compared to the previous look & feel it looks less professional, a bit more playful and over-colorized.
- Both of the above makes the UI "hit me in the eye", and less usable on a typical desktop. Everything a little smaller, less "candy-coloured". a little more decent colours and slightly less contrast should increase usability a lot, I feel.
- The Unicorn background has come back on the public pages, even though I have set a general override for background, default cover and browser icon in Administration → Custom Branding. This should apply to all stations, if nothing else is set there. (When using station branding, the uploaded background image is also ignored.)
- The public schedule page doesn’t show the "Playlist/Streamer" tooltips anymore.
- On the public schedule page, the heavy contrasts make the red arrow and timeline indicators almost invisible.

Unfortunately, I haven’t kept old screenshots for a side-by-side comparison, but here is a typical session on my laptop (1920x1080 screen). I typically have my browser only half to two-thirds wide, in order to be able to see other windows, like the console, a file manager or the like.

![Arbeitsfläche 1_014](https://github.com/AzuraCast/AzuraCast/assets/3706922/9643a1d1-b090-408c-98d8-129f6425ab76)

![Nite Radio – Mozilla Firefox_001](https://github.com/AzuraCast/AzuraCast/assets/3706922/4c9eff3d-8d50-46d0-b558-7eb3b73ba841)

![Schedule - Nite Radio – Mozilla Firefox_001](https://github.com/AzuraCast/AzuraCast/assets/3706922/1515582b-ed3a-4924-9fa5-aa6e3b0804d0)

Don’t take my points too harsh—I am aware that this is a work in progress. Just pointing out—at an early stage—what might affect users when using wonderful AzuraCast in the most efficient manner. Using the admin UI usually means _work_, and nothing should disturb a user when trying to work efficiently. Usability is key, and ease of use a great benefit.

**Comment by Vaalyn** [comment]
👋  Hey, thank you for the feedback, this was indeed quite a humongous task since due to the old components / framework that we used not being production ready / still being unfinished basically the whole UI needed to be reworked.

I'm currently looking at some of your findings to see how to address them.

> Fonts are bigger now, you can’t get as much onto a 2/3 wide browser on a 1920x1080 desktop.

The font sizes were increased a bit to reduce eye strain and to help people with visual impairments with readability.
This does indeed reduce the information density but I think this is a worthwhile tradeoff.

> Everything has much more contrasting colors, compared to the previous look & feel it looks less professional, a bit more playful and over-colorized.

- The blue is still the same hex color as before
- Adjusted the body background to the same grey we used previously to reduce the "in your face" effect that the near white caused
- Adjusted the card background colors (for light & dark theme slightly)
- Adjusted the general text color to match the dark grey we used previously
- The red and greens were breaking accessibility guidelines for contrast rates so they got darkened a bit
- The biggest differentiator is probably that the buttons on the profile page are full buttons instead of link buttons so I've changed this around to make them colored link buttons again

Here is a screenshot of how the profile page looks like after those changes:

![image](https://github.com/AzuraCast/AzuraCast/assets/13745863/e9148147-47b7-4fcb-acb2-472213351fbb)
![image](https://github.com/AzuraCast/AzuraCast/assets/13745863/98f02e70-1362-467e-a6b0-9b63fc2a0fd2)

> The Unicorn background has come back on the public pages, even though I have set a general override for background, default cover and browser icon in Administration → Custom Branding. This should apply to all stations, if nothing else is set there. (When using station branding, the uploaded background image is also ignored.)

This seems to be a bug 🤔

@Moonbase59 could you create a bug issue about this?

> The public schedule page doesn’t show the "Playlist/Streamer" tooltips anymore.

Looks like this was missed in the rework and needs to be implemented in the new UI.

> On the public schedule page, the heavy contrasts make the red arrow and timeline indicators almost invisible.

The red is still the same hex value: `#FF0000`
The general contrast seems to be much better in the new version for me:

**Before:** 
<img width="532" alt="image" src="https://github.com/AzuraCast/AzuraCast/assets/13745863/512eb674-9c80-453f-9b7a-61ed0310d284">

**After:**
<img width="463" alt="image" src="https://github.com/AzuraCast/AzuraCast/assets/13745863/ce8e9247-99f9-4c7b-8fd2-786767799c85">

**Comment by Moonbase59** [comment]
Hmm. I agree that one should keep people with disabilities in mind (like with the contrast, size and ARIA things). Must have been quite some work, getting rid of the old stuff and replacing it all with more modern components.

I still think
- the contrast is a little too much.
- a red button with white text on a blue background simply … hurts in the eyes
- things are not yet consistent overall:
  - some fonts (now playing) are rather small, while the card titles, for instance, hit you bare in the face
  - Some things that look like buttons (but are selection lists really) are black. Confusing.
- Overall (see my profile page screenshot) I think we’re wasting valuable screen estate. Imagine someone using it on a regular 24" or even 32" screen…
- It’s fortunately no problem nowadays to `Ctrl`+`Mousewheel` in order to make things bigger, or set a default zoom (for instance, my eyesight isn’t as good anymore, and I’ve set a Linux-wide 120% so window bars, menus, everything appear a little larger).
- I agree on the red time indicator in schedule, but ir’s hard to find if you have many scheduled playlists (see my screenshot—almost all days are "blue").

I can’t remember which tools I used back in the days when I did websites, but there were some that would display web pages like a colorblind person sees them (various kinds of colorblindness), and others to evaluate contrast and color usability. Maybe one could run the AzuraCast pages through those, to evaluate that a little more?

And/or also ask other AzuraCast users what they think? (Especially those with visual deficiencies.)

Anyway, function first (good job!), polish later… Thanks for coming back with your thoughts!

**Reply by BusterNeece** [reply]
I don't expect we're going to agree on matters of taste, and a lot of this simply boils down to that.

We've tried in as many places as possible to replicate the existing theme style, and I think we've done a very admirable job of that overall, while greatly reducing our browser JS and CSS load and bringing us up to modern tooling.

We pushed a few changes to resize things a little (and to fix the bug with custom backgrounds not working), but otherwise I'm satisfied with the work that's gone into this, which has indeed been over a week of pretty hard work.

**Comment by stickbear2015** [comment]
I know you guys have been working on what seems to be visual aspects for those disabled users who still have some sight and don't use a screen reader but Just a note, watch your labeling, straight buttons are fine, but when you start incorporating buttons into links, this can sometimes explode, and screen reader users can have a difficult time finding things they used to be able to find.
I don't expect this to be an issue because you all have been rockin' the accessibility, but I had to say it, and I don't have access to a non-production environment to run the usual tests to ensure things stay put during the re-write. Since your re-writing the world, can I beg and pleed that you fix the selection boxes in the media manager so they're actual checkboxes? It slows down my productivity if I have to find the select thing and hit it, sure it changes, and reads, but if I could just use my screen readers command to bounce between checkboxes, my life would be so much more productive, and I bet others would love me and yall to. :-) I mean if you can't I'll deal.

**Reply by BusterNeece** [reply]
@stickbear2015 Of course! In fact, as part of this styling update we've moved several buttons that were `a` tags to be actual `button` tags, now that our styling permits that without issue.

Accessibility is and always will be a huge priority for us, so if you run into issues with our new layout please let us know.

**Reply by BusterNeece** [reply]
@stickbear2015 Also, to your point, the new library we use for the Datatable component (used in the Media Manager and elsewhere) _does_ use native checkboxes for that functionality. We've long wanted to change that, and now we have. Hopefully that helps your experience.

**Comment by Moonbase59** [comment]
> We've tried in as many places as possible to replicate the existing theme style, and I think we've done a very admirable job of that overall, while greatly reducing our browser JS and CSS load and bringing us up to modern tooling.

That can’t be denied and is greatly appreciated, especially all the technical aspects behind it all.

I think @Vaalyn’s suggestions shown above are looking much better.

Keep up doing the good stuff!

**Comment by Moonbase59** [comment]
More feedback on the technical side, for Rolling Release #[0a46ef2](https://github.com/AzuraCast/AzuraCast/commit/0a46ef2a2f9d038774280e911e832da6b98486ff) (2023-07-09 14:55):

- Generally, it seems to consume less CPU: For 3 audio and two different video streams I can now get by with a load average of ~1.5, if nothing else is to do.
- CPU load increases heavily to a load average of ~3.5–4.0 when using SFTP to upload video files (usually 2–4 GB in size).
- Memory usage has increased by roughly 1 GB. I usually had ~2 GB RAM used, now it’s typically 3–3.2 GB, leaving not too much for caching on this 4 GB RAM VM.

**Reply by Vaalyn** [reply]
That's rather weird 🤔 

The LS team and I looked into the RAM usage over the past week so that we can optimize that part quite a lot.
They implemented "dead code elimination" so that only the parts of LS that are actually used will be loaded.

In my tests my stations with optimizations in the station profile disabled went from ~400mb per LS process down to around ~215mb and with optimizations for reduced RAM usage went from ~300mb down to ~185mb.

**Reply by Moonbase59** [reply]
That’s interesting, let’s keep an eye on it. I followed Liquidsoap’s efforts to make starting faster, balance CPU/RAM, and all the garbage collection stuff. Btw, it seems they phase out jemalloc again, we might eventually get rid of installing that again.

In my experience, over time, it seems to reduce memory consumption again, to ~2.5–2.8 GB in my case. I run AzuraCast on a Proxmox VM, 4 CPUs, 4 GB RAM, 80 GB disc, media NFS-mounted (~2 TB, ~165k songs, 59 playlists.

I generate

- a 128 kbit/s MP3 audio stream,
- a 128 kbit/s AAC audio stream,
- an HLS audio stream,
- a max 8,000 kbit/s 720p25 video stream (watched video folder + radio audio + overlays & API usage)
- a max 8.000 kbit/s 720p25 video stream with a static test image and a 1 kHz sine at -23 dB

and I’m rather happy with the current CPU load (load average ~1.5–2 during normal streaming operation, up to almost 4 when using sftpgo or media indexing).

Currently it doesn’t look critical, just something to watch out for, in case media gets uploaded (sftpgo uses _lots_ of CPU) or re-indexed.

---

Let’s see if anything changes—I’m currently test-driving Liquidsoap 2.2.0+git@0a90fcf99 in the AzuraCast docker, for some video features I need.

**Reply by Vaalyn** [reply]
Maybe it has something to do with having video streams, I didn't have any in my tests so that could be possible 🤔 

How much is a single LS process using for you?
Could you check what difference there is when you disable the video streams?

**Reply by Moonbase59** [reply]
Can take a mo, just debugging something in LS.

> How much is a single LS process using for you?

You mean one instance of LS, including all threads? Say, using `htop`?

> Could you check what difference there is when you disable the video streams?

Yup, no problem. I made it so they can all be added at the end of LS config, so removing will be easy.

**Reply by Moonbase59** [reply]
Removed all video code for a test now. AzuraCast display shows 1,9 GB RAM + short of 1 GB cached, htop 1.16 GB. Load averages between 0.3 and 1 during normal streaming (no sftpgo, no MediaCheck and the like). Htop shows LS CPU% between 13 and 15, CheckMediaTask sometimes hitting in with up to 100%.

**Reply by Vaalyn** [reply]
> You mean one instance of LS, including all threads? Say, using htop?

Yeah, see my reference screenshot from my test server where I run a few stations cloned from my prod station.
There I have different combinations of optimizations and settings running to check multiple scenarios.

If you set htop to tree mode it's more easy to see which is the main process of LS, take the RAM usage from that:

<img width="1110" alt="image" src="https://github.com/AzuraCast/AzuraCast/assets/13745863/845512ec-1816-4a19-943b-3fbfe6dba5bb">

**Reply by Moonbase59** [reply]
Yup, here’s my `htop` with threads shown and tree view. All looks easy going… I wonder where the high RAM usage display in AzuraCast comes from? `htop` shows much less.

![matthias@radio: ~_004](https://github.com/AzuraCast/AzuraCast/assets/3706922/8f89a9b3-ee52-4d03-9c0d-a6e10d22a4e3)

![Auswahl_232](https://github.com/AzuraCast/AzuraCast/assets/3706922/e9f85dd4-1e0e-4b0a-ae24-f0b9cffbde63)

**Reply by Vaalyn** [reply]
Is that screenshot with or without video streaming enabled?
Any of the optimization options in profile -> autodj enabled?

>  I wonder where the high RAM usage display in AzuraCast comes from? htop shows much less.

That is quite a weird difference since use the same source for the raw ram usage data that htop uses, namely `/proc/meminfo`

There is also not really much going on when preparing that data:
- https://github.com/AzuraCast/AzuraCast/blob/main/src/Service/MemoryStats.php
- https://github.com/AzuraCast/AzuraCast/blob/main/src/Service/MemoryStats/MemoryData.php

So there might be something that htop calculates based on the raw data that we might not be doing? 🤔

**Reply by Moonbase59** [reply]
I wonder… Yeah, video code completely removed from LS config, and restarted broadcasting. Btw, I’m currently running LS 2.2.0+git@0a90fcf99 in this docker, but RAM-wise it is exactly the same, just verified.

No optimiziations selected, I always try to run without.

Re `htop`: Did you run it outside the docker, like I did?

All in all, nothing to get crazy about, I think. I’ve had an (AzuraCast-shown) RAM usage of 1.8–2.2 GB for a long time (including the video), going back to versions long before the Bootstrap 5/Vue 3 change. I’m actually rather delighted that CPU usage seems to decrease with every new version of Liquidsoap and AzuraCast, doing the same jobs.

It’s just that sudden increase to more than 3 GB shown in the latest versions I was wondering about. Maybe `htop` is right, after all.

If it helps, here’s my `/proc/meminfo` from inside the docker:
```
MemTotal:        4005344 kB
MemFree:          875588 kB
MemAvailable:    2411572 kB
Buffers:           61424 kB
Cached:          1211308 kB
SwapCached:        21156 kB
Active:           877536 kB
Inactive:        1257664 kB
Active(anon):     237828 kB
Inactive(anon):   738236 kB
Active(file):     639708 kB
Inactive(file):   519428 kB
Unevictable:       27768 kB
Mlocked:           27768 kB
SwapTotal:       4017148 kB
SwapFree:        3455632 kB
Dirty:              1584 kB
Writeback:             0 kB
AnonPages:        878268 kB
Mapped:           362876 kB
Shmem:            104836 kB
KReclaimable:     665932 kB
Slab:             785164 kB
SReclaimable:     665932 kB
SUnreclaim:       119232 kB
KernelStack:       46544 kB
PageTables:        49984 kB
NFS_Unstable:          0 kB
Bounce:                0 kB
WritebackTmp:          0 kB
CommitLimit:     6019820 kB
Committed_AS:   35799824 kB
VmallocTotal:   34359738367 kB
VmallocUsed:       65060 kB
VmallocChunk:          0 kB
Percpu:             2544 kB
HardwareCorrupted:     0 kB
AnonHugePages:      2048 kB
ShmemHugePages:        0 kB
ShmemPmdMapped:        0 kB
FileHugePages:         0 kB
FilePmdMapped:         0 kB
HugePages_Total:       0
HugePages_Free:        0
HugePages_Rsvd:        0
HugePages_Surp:        0
Hugepagesize:       2048 kB
Hugetlb:               0 kB
DirectMap4k:      309096 kB
DirectMap2M:     3885056 kB
```

**Reply by Vaalyn** [reply]
> Re htop: Did you run it outside the docker, like I did?

No, I ran it inside the container but just checked from outside and it's no different than the htop inside the container.

**Reply by Moonbase59** [reply]
In https://github.com/AzuraCast/AzuraCast/blob/main/src/Service/MemoryStats/MemoryData.php, I miss the "buffers" part. On Linux, I’d usually calculate

```
used = total - free - buffers - cached
```

This is what `top` and `htop` seem to be doing. Then again, the few buffers shouldn’t make that much difference, at least not in my example. You still might

```php
<?php

declare(strict_types=1);

namespace App\Service\MemoryStats;

use App\Radio\Quota;
use Brick\Math\BigInteger;

final class MemoryData
{
    public function __construct(
        public readonly BigInteger $memTotal,
        public readonly BigInteger $memFree,
        public readonly BigInteger $buffers,
        public readonly BigInteger $cached,
        public readonly BigInteger $swapTotal,
        public readonly BigInteger $swapFree,
    ) {
    }

    public static function fromMeminfo(array $meminfo): self
    {
        $memTotal = Quota::convertFromReadableSize($meminfo['MemTotal']) ?? BigInteger::zero();
        $memFree = Quota::convertFromReadableSize($meminfo['MemFree']) ?? BigInteger::zero();
        $buffers = Quota::convertFromReadableSize($meminfo['Buffers']) ?? BigInteger::zero();
        $cached = Quota::convertFromReadableSize($meminfo['Cached']) ?? BigInteger::zero();
        $swapTotal = Quota::convertFromReadableSize($meminfo['SwapTotal']) ?? BigInteger::zero();
        $swapFree = Quota::convertFromReadableSize($meminfo['SwapFree']) ?? BigInteger::zero();

        return new self(
            $memTotal,
            $memFree,
            $buffers,
            $cached,
            $swapTotal,
            $swapFree
        );
    }

    public function getUsedMemory(): BigInteger
    {
        return $this->memTotal
            ->minus($this->memFree)
            ->minus($this->buffers)
            ->minus($this->cached);
    }

    public function getUsedSwap(): BigInteger
    {
        return $this->swapTotal->minus($this->swapFree);
    }
}
```

Buffers and Cache are usually shown together, for instance in `free`, `top` and others, so probably the yellow "Cached" bar should also be a "Buffers/Cached" value.

**Reply by Vaalyn** [reply]
Yeah, thought the same so gonna add that to the code, but I still wonder were the remaining difference between the value we calculate and htop then comes from... 🤔 

Based on the current values I have seen on my test installation there is still a difference of ~200mb between htop and the calculation from us:

<img width="1110" alt="image" src="https://github.com/AzuraCast/AzuraCast/assets/13745863/65c9eb94-1faf-4055-bd48-66f3ec7ee055">

I'll put the output of `/proc/meminfo` for that screenshot here.
If you see anything that could explain the remaining difference I'll gladly add that to the calculation code too.

<details> 
  <summary>/proc/meminfo output</summary>

  ```
  MemTotal:        8272692 kB
  MemFree:          224276 kB
  MemAvailable:    5773256 kB
  Buffers:          235680 kB
  Cached:          5455920 kB
  SwapCached:         3280 kB
  Active:          5189612 kB
  Inactive:        2207228 kB
  Active(anon):    1518632 kB
  Inactive(anon):   257400 kB
  Active(file):    3670980 kB
  Inactive(file):  1949828 kB
  Unevictable:       18600 kB
  Mlocked:           18600 kB
  SwapTotal:       4030460 kB
  SwapFree:        3991804 kB
  Dirty:              3816 kB
  Writeback:             0 kB
  AnonPages:       1722656 kB
  Mapped:           432024 kB
  Shmem:             69032 kB
  KReclaimable:     242856 kB
  Slab:             418480 kB
  SReclaimable:     242856 kB
  SUnreclaim:       175624 kB
  KernelStack:       16448 kB
  PageTables:        22068 kB
  NFS_Unstable:          0 kB
  Bounce:                0 kB
  WritebackTmp:          0 kB
  CommitLimit:     8166804 kB
  Committed_AS:   11369128 kB
  VmallocTotal:   34359738367 kB
  VmallocUsed:       39180 kB
  VmallocChunk:          0 kB
  Percpu:             7488 kB
  HardwareCorrupted:     0 kB
  AnonHugePages:      6144 kB
  ShmemHugePages:        0 kB
  ShmemPmdMapped:        0 kB
  FileHugePages:         0 kB
  FilePmdMapped:         0 kB
  CmaTotal:              0 kB
  CmaFree:               0 kB
  HugePages_Total:       0
  HugePages_Free:        0
  HugePages_Rsvd:        0
  HugePages_Surp:        0
  Hugepagesize:       2048 kB
  Hugetlb:               0 kB
  DirectMap4k:      438120 kB
  DirectMap2M:     7950336 kB
  DirectMap1G:           0 kB
  ```
</details>

**Reply by Moonbase59** [reply]
We _are_ using "real" (1024-based) conversion & display in AzuraCast, right?

Ever since the SI units crap (1000-based) was forced on us, we see "wrong" indicators (as in `/proc/meminfo`, which is not "kB" but "kiB" really). `free` uses "kiB", `htop` has "GiB" (I think), and `top` uses "MiB". A mess.

Still I don’t see where the difference comes from—seems not to be a GB/GiB thing.

**Reply by Moonbase59** [reply]
I _am_ using my code above now, but I don’t see much difference:

![Auswahl_233](https://github.com/AzuraCast/AzuraCast/assets/3706922/bd731285-b563-4732-a962-156ae87817b1)

Maybe the memory consumption is really due to having –165,000 songs and 59 playlists…

**Reply by Vaalyn** [reply]
@Moonbase59 I've done some adjustments on the calculation, see https://github.com/AzuraCast/AzuraCast/pull/6412

**Reply by Moonbase59** [reply]
@Vaalyn See https://github.com/AzuraCast/AzuraCast/pull/6412#issuecomment-1629020135

**Comment by Moonbase59** [comment]
More feedback: I quite like the new possibility to consciously choose between Reload and Restart when you make changes and AzuraCast prompts for "Reload to apply changes". Very well done, and sometimes really useful!

**Comment by Moonbase59** [comment]
Even more feedback! We see what @BusterNeece & @Vaalyn are working on, and they have taken a massive amount of work on their shoulders, changing the whole frontend code, and getting rid of bugs & leftovers.

But I strongly feel they’re on a good track with that decision—a little bumpiness here and there is certainly worth it.

:heart: for what they’re giving us, and for all the hard work!
